src/engine/strat_engine/serde_structs.rs

```diff
@@ -12,12 +12,61 @@
 // can convert to or from them when saving our current state, or
 // restoring state from saved metadata.
 
-use serde::Serialize;
+use serde::{Serialize, Serializer};
 
 use devicemapper::{Sectors, ThinDevId};
 
 use crate::engine::types::{DevUuid, FilesystemUuid};
 
+const MAXIMUM_STRING_SIZE: usize = 255;
+
+// Find the largest index which occurs on a char boundary of value which is no
+// greater than len.
+// TODO: Replace this method with String::floor_char_boundary when
+// possible.
+fn our_floor_char_boundary(value: &str, len: usize) -> usize {
+    let len = std::cmp::min(len, value.len());
+
+    let mut new_index = len;
+    while !value.is_char_boundary(new_index) && new_index != 0 {
+        new_index -= 1;
+    }
+
+    new_index
+}
+
+// Return a new String, split at the highest index which lies on a char
+// boundary of value which is no greater than len.
+fn safe_split_at(value: &str, len: usize) -> String {
+    value
+        .split_at(our_floor_char_boundary(value, len))
+        .0
+        .to_string()
+}
```

Comment:
One last comment on this. This doesn't avoid the unnecessary allocation which we can do using a slice index and that will return a `&str` which we can then serialize without an intermediary allocation. It shouldn't panic because we've checked the char boundary.